File: attack_phase/washington-v1.1-release/src/car/src/uart.c
#include <stdint.h>
#include <stdbool.h>
#include <stddef.h>

#include "driverlib/sysctl.h"
#include "driverlib/gpio.h"
#include "driverlib/pin_map.h"
#include "driverlib/uart.h"
#include "driverlib/eeprom.h"
#include "inc/hw_memmap.h"
#include "uart.h"
#include "util.h"

/**
 * @brief Initializes device <-> device and host <-> device UART lines
 * 
 */
void uart_init(void) {

    //Enable Device <--> Device UART
    SysCtlPeripheralEnable(SYSCTL_PERIPH_UART1);
    SysCtlPeripheralEnable(SYSCTL_PERIPH_GPIOB);

    GPIOPinConfigure(GPIO_PB0_U1RX);
    GPIOPinConfigure(GPIO_PB1_U1TX);

    GPIOPinTypeUART(GPIO_PORTB_BASE, GPIO_PIN_0 | GPIO_PIN_1);

    // Configure the UART for 115,200, 8-N-1 operation.
    UARTConfigSetExpClk(
        (uint32_t)UART1_BASE, SysCtlClockGet(), 115200,
        (UART_CONFIG_WLEN_8 | UART_CONFIG_STOP_ONE | UART_CONFIG_PAR_NONE));

    /*********************************************************************/

    //Enable Computer <--> Device UART
    SysCtlPeripheralEnable(SYSCTL_PERIPH_UART0);
    SysCtlPeripheralEnable(SYSCTL_PERIPH_GPIOA);

    GPIOPinConfigure(GPIO_PA0_U0RX);
    GPIOPinConfigure(GPIO_PA1_U0TX);

    GPIOPinTypeUART(GPIO_PORTA_BASE, GPIO_PIN_0 | GPIO_PIN_1);

    // Configure the UART for 115,200, 8-N-1 operation.
    UARTConfigSetExpClk(
        (uint32_t)UART0_BASE, SysCtlClockGet(), 115200,
        (UART_CONFIG_WLEN_8 | UART_CONFIG_STOP_ONE | UART_CONFIG_PAR_NONE));
    
    /*********************************************************************/

    //Flush garbage data
    while (UARTCharsAvail((uint32_t)UART1_BASE)) {
        UARTCharGet((uint32_t)UART1_BASE);
    }
    while (UARTCharsAvail((uint32_t)UART0_BASE)) {
        UARTCharGet((uint32_t)UART0_BASE);
    }
}


/**
 * @brief Sends a message over a given UART port
 * 
 * As of right now, the entirety of the struct is sent over.
 * Perhaps in the future we could send each field individually, but this seems to work just fine.
 * 
 * @param PORT the UART Port to use
 * @param message the message to send
 */
void uart_send_message(const uint32_t PORT, Message* message) {

    for(size_t i = 0; i < 4; i++) {
        UARTCharPut(PORT, uart_magic[i]);
    }

    uint8_t* struct_ptr = (uint8_t*) message;

    //send everything in message
    for(size_t i = 0; i < sizeof(Message); i++) {

        while(UARTBusy(PORT));

        #ifdef DEBUG
        UARTCharPut(HOST_UART, struct_ptr[i]);
        #endif

        UARTCharPut(PORT, struct_ptr[i]);

        if(i % 8 == 0) {
            UARTCharGet(PORT);
        }

    }
}

/**
 * @brief Sends raw bytes over a given UART PORT
 * 
 * @param PORT the UART Port to use
 * @param message the buffer to send
 * @param size the number of bytes to send
 */
void uart_send_raw(const uint32_t PORT, void* message, size_t size) {
    for(size_t i = 0; i < size; i++) {
        UARTCharPut(PORT, ((uint8_t*) message)[i]);
    }
}

void uart_read_raw(const uint32_t PORT, void* message, size_t size) {
    for(size_t i = 0; i < size; i++) {
        ((uint8_t*) message)[i] = UARTCharGet(PORT);
    }
}

//Possible error: apparently uart buffer size is only 16 bytes
/**
 * @brief Attempts to read a message struct into a given buffer.
 * 
 * [NOTE]: This function offers no integrity verification!!! Corruption is definitely possible.
 * In the future, the UART buffer should be periodically flushed until a magic is found (resolved?)
 * 
 * @param PORT the UART Port to use
 * @param message the buffer to write to
 * @return true if the buffer was filled up
 * @return false if the read timed out
 */
bool uart_read_message(const uint32_t PORT, Message* message) {
    size_t i = 0;
    size_t j = 0;
    size_t timeout = 0; //really not a fan of doing timeouts like these... maybe we should use interrupts?

    #define TIMEOUT_THRESHOLD 100000 //not necessary, but might as well leave it in i guess


    //go through uart buffer until you find the magic header "0ops"
   while(j < 4) {

        if(!UARTCharsAvail(PORT)) {
            timeout++;
            if(timeout > TIMEOUT_THRESHOLD) {
                #ifdef DEBUG
                debug_print("[d] no magic\n");
                #endif
                return false;
            }
        }

        else {
             uint8_t c = UARTCharGet(PORT);
             #ifdef DEBUG
                uart_send_raw(HOST_UART, &c, sizeof(c));
            #endif

            if(uart_magic[j] != c) {
                timeout = 0;
                j = 0;
            }
             else {
                timeout = 0;
                j++;
            }
        }
    }

    timeout = 0;

    #ifdef DEBUG
    debug_print("Recieved Message: [");
    #endif

    while(i < sizeof(Message)) {

        if(!UARTCharsAvail(PORT)) {
            continue;
        }

        uint8_t c = UARTCharGetNonBlocking(PORT);

        ((uint8_t*) message)[i] = c;

        #ifdef DEBUG
        uart_send_raw(HOST_UART, &c, sizeof(c));
        #endif
        
        if(i % 8 == 0) {
            UARTCharPut(PORT, 'A');
        }

        i++;
    }

    #ifdef DEBUG
    debug_print("] END |");
    #endif

    return true;
}

/**
 * @brief Initializes the internal eeprom. Resets the board on failure.
 * 
 */
void eeprom_init(void) {
    SysCtlPeripheralEnable(SYSCTL_PERIPH_EEPROM0);

    //if eeprom initialization fails, reset
    if(EEPROMInit() != EEPROM_INIT_OK) {
        #ifdef DEBUG
        debug_print("eeprom load failure\n");
        #endif
        SysCtlReset();
    }
}

//Not sure if EEPROM is mapped as an iommu device - do some research and see if attacks that way are possible?

//TODO: Add boundry checks to read/write methods
/**
 * @brief Wrapper to read len bytes into a buffer from a given address
 * 
 * @param msg buffer to read into
 * @param len number of bytes to read
 * @param address read address (starts at 0)
 */
void eeprom_read(void* msg, size_t len, size_t address) {
    EEPROMRead(msg, address, len);
}


/**
 * @brief Wrapper to write len bytes from a buffer to a given EEPROM address
 * 
 * @param msg buffer to write from
 * @param len number of bytes to write
 * @param address write address (starts at 0)
 */
void eeprom_write(void* msg, size_t len, size_t address) {
    EEPROMProgram(msg, address, len);
}

File: attack_phase/wpi-v1.2-release/wpi-v1.2-release/src/fob/src/hw/uart.rs
//! UART communication support

#![allow(dead_code)] // Remove this for actual use, just so the compiler doesn't complain for the
                     // hello world.
use cortex_m::singleton;

use crate::tivaware::{
    GPIOPadConfigSet, GPIOPinConfigure, GPIOPinTypeUART, SysCtlClockGet, SysCtlPeripheralEnable,
    UARTCharGet, UARTCharPut, UARTCharsAvail, UARTConfigSetExpClk, UARTFIFOLevelSet, GPIO_PA0_U0RX,
    GPIO_PA1_U0TX, GPIO_PIN_0, GPIO_PIN_1, GPIO_PIN_TYPE_STD_WPU, GPIO_PORTA_BASE,
    GPIO_PORTB_BASE, GPIO_STRENGTH_2MA, SYSCTL_PERIPH_GPIOA, SYSCTL_PERIPH_GPIOB,
    SYSCTL_PERIPH_UART0, SYSCTL_PERIPH_UART1, UART0_BASE, UART1_BASE, UART_CONFIG_PAR_NONE,
    UART_CONFIG_STOP_ONE, UART_CONFIG_WLEN_8, UART_FIFO_RX1_8, UART_FIFO_TX1_8, UARTFIFOEnable, GPIO_PB0_U1RX, GPIO_PB1_U1TX,
};

use crate::comms::io::IO;
use crate::comms::packet::PacketIO;
use crate::comms::io::Error;

/// A UART communication channel.
pub trait UartIO: IO {
    /// The base address of the UART peripheral's registers.
    const UART_BASE: u32;
}

impl<T: UartIO> IO for T {
    fn data_avail(&self) -> bool {
        unsafe { UARTCharsAvail(Self::UART_BASE) }
    }

    fn write_byte(&self, byte: u8) {
        unsafe { UARTCharPut(Self::UART_BASE, byte) }
    }

    fn read_byte(&self) -> u8 {
        unsafe { UARTCharGet(Self::UART_BASE) as u8 }
    }

    fn read_byte_timeout<'a, 'b: 'a>(&'a self, time_keeper: &'b dyn crate::utils::timing::Timeout) -> Result<u8, Error> {
        loop {
            if time_keeper.has_expired() {
                return Err(Error::TimedOut);
            }

            if self.data_avail() {
                return Ok(self.read_byte());
            }
        }
    }
}

/// Board<->Host UART communication channel.
pub struct HostUart;
impl HostUart {
    pub fn new() -> Option<&'static mut Self> {
        unsafe {
            SysCtlPeripheralEnable(SYSCTL_PERIPH_UART0);
            SysCtlPeripheralEnable(SYSCTL_PERIPH_GPIOA);

            GPIOPinConfigure(GPIO_PA0_U0RX);
            GPIOPinConfigure(GPIO_PA1_U0TX);

            GPIOPinTypeUART(GPIO_PORTA_BASE, (GPIO_PIN_0 as u8) | (GPIO_PIN_1 as u8));
            GPIOPadConfigSet(
                GPIO_PORTA_BASE,
                (GPIO_PIN_0 as u8) | (GPIO_PIN_1 as u8),
                GPIO_STRENGTH_2MA,
                GPIO_PIN_TYPE_STD_WPU,
            );

            UARTConfigSetExpClk(
                UART0_BASE,
                SysCtlClockGet(),
                115200,
                UART_CONFIG_WLEN_8 | UART_CONFIG_STOP_ONE | UART_CONFIG_PAR_NONE,
            );
            UARTFIFOLevelSet(UART0_BASE, UART_FIFO_TX1_8, UART_FIFO_RX1_8);
            UARTFIFOEnable(UART0_BASE);
        };

        singleton!(: HostUart = HostUart {})
    }
}

impl UartIO for HostUart {
    const UART_BASE: u32 = UART0_BASE;
}

impl PacketIO for HostUart {}

/// Board<->Board UART communication channel.
pub struct BridgeUart;
impl BridgeUart {
    pub fn new() -> Option<&'static mut Self> {
        unsafe {
            SysCtlPeripheralEnable(SYSCTL_PERIPH_UART1);
            SysCtlPeripheralEnable(SYSCTL_PERIPH_GPIOB);

            GPIOPinConfigure(GPIO_PB0_U1RX);
            GPIOPinConfigure(GPIO_PB1_U1TX);

            GPIOPinTypeUART(GPIO_PORTB_BASE, (GPIO_PIN_0 as u8) | (GPIO_PIN_1 as u8));
            GPIOPadConfigSet(GPIO_PORTB_BASE, (GPIO_PIN_0 as u8) | (GPIO_PIN_1 as u8), GPIO_STRENGTH_2MA, GPIO_PIN_TYPE_STD_WPU);

            UARTConfigSetExpClk(
                UART1_BASE,
                SysCtlClockGet(),
                115200,
                UART_CONFIG_WLEN_8 | UART_CONFIG_STOP_ONE | UART_CONFIG_PAR_NONE,
            );
            UARTFIFOLevelSet(UART1_BASE, UART_FIFO_TX1_8, UART_FIFO_RX1_8);
            UARTFIFOEnable(UART1_BASE);
        };

        singleton!(: BridgeUart = BridgeUart {})
    }
}

impl UartIO for BridgeUart {
    const UART_BASE: u32 = UART1_BASE;
}

impl PacketIO for BridgeUart {}
